crates/typst/src/lint.rs

```diff
@@ -0,0 +1,44 @@
+use ecow::EcoString;
+
+use crate::{
+    diag::{warning, SourceDiagnostic},
+    syntax::{ast::Strong, Source, SyntaxKind, SyntaxNode},
+};
+
+/// Lints a [`Source`] and returns a list of all warnings found.
+pub fn lint(source: &Source) -> Vec<SourceDiagnostic> {
+    let mut warnings = Vec::new();
+    let mut parents = vec![source.root()];
+
+    while let Some(node) = parents.pop() {
+        parents.extend(node.children().rev());
+
+        warnings.extend(lint_node(node));
+    }
```

Comment:
any reason to maintain an depth-first search stack instead of doing this recursively?

Reply:
It was done this way in the original PR, no reason it can't be changed. Would doing it recursively improve efficiency?

Reply:
I would save the allocation. Probably doesn't make much of a difference in practice.

Reply:
Makes sense. Should either be changed to use a proper queue to avoid unnecessary re-allocations, or done recursively for efficiency. I think the stackoverflow risk is low/non-existent, so I'm leaning towards the second option.

Reply:
I would write it in the recursive way. The interpreter itself is also recursive, so if this overflows the stack, it's even way worse for the interpreter because it uses more stack space per node.

Reply:
Luckily I just updated the PR so this loop no longer exists 🥲